File: leetcode/minstack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.min = []


    def push(self, val: int) -> None:
        self.stack.append(val)

        if len(self.min) == 0:
            self.min.append(val)
            return

        if val <= self.min[-1]:
            self.min.append(val)

    def pop(self) -> None:
        if self.stack[-1] == self.min[-1]:
            self.min.pop(-1)
        self.stack.pop(-1)


    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: leetcode/minstack.cpp
#include <stack>

class MinStack {
private:
    std::stack<int> min{};
    std::stack<int> stack{};
public:
    MinStack() {
        
    }
    
    void push(int val) {
        if (min.empty() || val <= min.top()) {
            min.push(val);
        }
        stack.push(val);
    }
    
    void pop() {
        if (!stack.empty() && !min.empty() && stack.top() == min.top()) {
            min.pop();
        }

        stack.pop();
    }
    
    int top() {
        return stack.empty() ? 0 : stack.top();
    }
    
    int getMin() {
        return min.empty() ? 0 : min.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */